### PR TITLE
Better prints with random params (issue #497)

### LINKED ISSUE #497: Storing, printing random parameters for netsim
@AdrienLeGuillou : in developing this example for netsim:

```r
## Example SIR model parameterization with fixed and random parameters
# Network model estimation
nw <- network_initialize(n = 100)
formation <- ~edges
target.stats <- 50
coef.diss <- dissolution_coefs(dissolution = ~offset(edges), duration = 20)
est <- netest(nw, formation, target.stats, coef.diss, verbose = FALSE)

# Random parameter list
my_randoms <- list(
  act.rate = param_random(1:3),
  inf.prob = function() rbeta(1, 1, 2)
)

# Parameters, initial conditions, and control settings
param <- param.net(rec.rate = 0.02, random.params = my_randoms)
init <- init.net(i.num = 10, r.num = 0)
control <- control.net(type = "SIR", nsteps = 100, nsims = 1,
                       resimulate.network = TRUE, tergmLite = TRUE)

# Simulate the model
sim <- netsim(est, param, init, control)

# Print and plot
sim
```

The print output is:
```r
EpiModel Simulation
=======================
Model class: netsim

Simulation Summary
-----------------------
Model type: SIR
No. simulations: 5
No. time steps: 100
No. NW groups: 1

Model Parameters
-----------------------
rec.rate = 0.02
random.params = <list>
act.rate = 1
inf.prob = 0.5108686

Model Output
-----------------------
Variables: s.num i.num r.num num si.flow ir.flow
Networks: sim1 ... sim5
Transmissions: sim1 ... sim5
```

The two random parameters, `act.rate` and `inf.prob`, shown above are (I believe) just the last drawn value of each (for simulation 5/5). This is misleading for the print output, plus I do not believe there is any way to look up the list of actual parameter values used in the simulation. Could you:

- [ ] Store the random parameter values, probably in a sublist of `dat$param`, unless you have a better suggestion
- [ ] Fix `print.netsim` so that it will not print the random parameter values. Maybe instead it should print the vector of selected values?
- [ ] Warning message if the random parameters would override anything in the default (non-random) parameter list

### PR DESCRIPTION
**Pull request by AdrienLeGuillou**
fixes #497 

- moves `generate_random_params` outside of the `withCallingHandlers` in
  `netsim_loop`
- store the randomly drawn values in `param$random.params.values`
- messages when random params overwrite existing ones.
- better print param:
    - single function to format the parameters
    - separate "fixed" params from "random" ones
    - print one instantiation of the random parameters

@smjenness do you prefer warnings instead of messages for parameter overwrite? If so, I will need to add a control to prevent them so the examples could pass the CRAN checks.

### CODE REVIEW

**Comment by smjenness** [comment]
> @smjenness do you prefer warnings instead of messages for parameter overwrite? If so, I will need to add a control to prevent them so the examples could pass the CRAN checks.

I think a warning would be fine. Why is a control needed? You can unit test that these properly generate a warning with testthat::expect_warning, right?

**Comment by AdrienLeGuillou** [comment]
the only problem with the warning is that we have to make sure that every test / example does not produce the warning or do it inside an `expect_warning` otherwise R CMD CHECK will convert it to an error. 

That's not a huge deal, and I suppose most people don't really want to see a warning if it is expected behavior. However, I also can see the other side of the argument, where a  warning will make it easier to spot it if not voluntary

**Comment by smjenness** [comment]
I guess I don't see the problem, because none of the existing tests use the random parameter API. And we wouldn't build new unit tests that would generate this warning (i.e., by use both a fixed and random parameter).

**Comment by AdrienLeGuillou** [comment]
Works for me! I will switch to warnings then

File: R/net.inputs.R [hunk 1]
```diff
@@ -354,12 +356,16 @@ param_random <- function(values, prob = NULL) {
 #' # Parameter model with deterministic and random parameters
 #' param <- param.net(inf.prob = 0.3, random.params = my_randoms)
 #'
+#' # A warning is produced if a random parameter overwrite a deterministic one
+#'
```
Comment (smjenness) [inline R/net.inputs.R:360]: @AdrienLeGuillou did you mean to include an example here? Also, let's edit the above text to say "...random parameter overwrites a fixed parameter."

Comment (AdrienLeGuillou) [reply R/net.inputs.R:360]: In this example, `acte.rate` is a random parameter but is also defined in the default `param.net` so a warning actually occurs. 

Maybe I should in this example use something else than `act.rate` to give an example without warnings and/or another one where I explicitly define a parameter as random and fixed

**Comment by AdrienLeGuillou** [comment]
Last push addresses the discussion on issue #497 

furthermore, `print.netsim` now directly call `print.param.net` internally. It makes the output more consistent and the code more maintainable.

**Review by smjenness — CHANGES_REQUESTED** [review]
See comments at bottom of #497

**Comment by AdrienLeGuillou** [comment]
I cleaned up the output,
this sample code can showcase the results (pasted below)

```r
## Example SIR model parameterization with fixed and random parameters
# Network model estimation
devtools::load_all()
nw <- network_initialize(n = 100)
formation <- ~edges
target.stats <- 50
coef.diss <- dissolution_coefs(dissolution = ~offset(edges), duration = 20)
est <- netest(nw, formation, target.stats, coef.diss, verbose = FALSE)

# Random parameter list
my_randoms <- list(
  act.rate = param_random(1:3),
  act.rate2 = function() runif(2),
  inf.prob = function() rbeta(1, 1, 2)
)

# Parameters, initial conditions, and control settings
param <- param.net(rec.rate = 0.02, rec.rate.2 = c(0.1, 0.2), random.params = my_randoms)
init <- init.net(i.num = 10, r.num = 0)
control <- control.net(type = "SIR", nsteps = 10, nsims = 20,
                       resimulate.network = FALSE, tergmLite = FALSE, ncore = 8)

# Print the params before simulation
print(param)

# Simulate the model
sim <- netsim(est, param, init, control)

# Print the simulation object
print(sim)
```

Print of the param object before netsim

```
Fixed Parameters
---------------------------
rec.rate = 0.02
rec.rate.2 = 0.1 0.2
act.rate = 1

Random Parameters
(Not drawn yet)
---------------------------
act.rate  = <function>
act.rate2  = <function>
inf.prob  = <function>
```

print of the netsim object
 
```
EpiModel Simulation
=======================
Model class: netsim

Simulation Summary
-----------------------
Model type: SIR
No. simulations: 20
No. time steps: 10
No. NW groups: 1

Fixed Parameters
---------------------------
rec.rate = 0.02
rec.rate.2 = 0.1 0.2
groups = 1

Random Parameters
---------------------------
act.rate = 2 3 3 2 3 1 2 1 1 2 ...
act.rate2 = <list>
inf.prob = 0.4717403 0.01566564 0.9718512 0.8246946 0.1272893 0.6503384
0.2097155 0.01845143 0.1150606 0.6180491 ...


Model Output
-----------------------
Variables: s.num i.num r.num num si.flow ir.flow
Networks: sim1 ... sim20
Transmissions: sim1 ... sim20

```

**Comment by AdrienLeGuillou** [comment]
@smjenness 
There is a test checking the output of `print(param)`

```r
test_that("print.param", {
  p <- param.dcm(inf.prob = 0.1, rec.rate = 0.1)
  expect_output(print(p), "DCM Parameters")

  p <- param.icm(inf.prob = 0.1, rec.rate = 0.1)
  expect_output(print(p), "ICM Parameters")

  p <- param.net(inf.prob = 0.1, rec.rate = 0.1)
  expect_output(print(p), "Network Model Parameters")
  expect_output(print(p), "act.rate = 1")
})
```

I can change them to get "Fixed Parameters" instead of "Network Model Parameters". Should I do this or keep "Network Model Parameters" and add "Fixed Parameters" and "Random Parameters" as sub-headers?

**Review by smjenness — APPROVED** [review]

**Comment by smjenness** [comment]
@AdrienLeGuillou : the output pre and post netsim looks good. You can just update the unit test to look for "Fixed Parameters" instead of "Network Model Parameters".

**Review by smjenness — APPROVED** [review]

**Comment by smjenness** [comment]
Great work @AdrienLeGuillou, thanks!!

**Pull request merged**